File: src/mobile/navigation_controller/src/human_detector.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <cv_bridge/cv_bridge.h>
#include <std_msgs/Float32.h>
#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>

class HumanDetector
{
private:
    ros::NodeHandle nh_;
    ros::Subscriber image_sub_;
    ros::Publisher human_area_pub_;
    cv::dnn::Net net_;  // YOLOv8 網路模型

public:
    HumanDetector()
    {
        // 訂閱影像
        image_sub_ = nh_.subscribe("/camera/image_raw", 1, &HumanDetector::imageCallback, this);
        
        // 設定發佈 topic
        human_area_pub_ = nh_.advertise<std_msgs::Float32>("/human_detection_area", 10);
        
        // 載入 YOLOv8 模型 (需下載 yolov8.onnx)
        net_ = cv::dnn::readNet("yolov8m.pt");

        // 使用 OpenCV DNN 加速
        net_.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
        net_.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);

        ROS_INFO("HumanDetector Node Initialized");
    }

    void imageCallback(const sensor_msgs::ImageConstPtr& msg)
    {
        try
        {
            // 轉換 ROS 影像到 OpenCV
            cv::Mat image = cv_bridge::toCvShare(msg, "bgr8")->image;
            cv::Mat blob = cv::dnn::blobFromImage(image, 1.0 / 255.0, cv::Size(640, 640), cv::Scalar(), true, false);

            // 設定網路輸入
            net_.setInput(blob);

            // 取得偵測結果
            std::vector<cv::Mat> outputs;
            net_.forward(outputs, net_.getUnconnectedOutLayersNames());

            float detected_human_area = 0.0;

            // 解析 YOLOv8 結果
            for (size_t i = 0; i < outputs.size(); i++)
            {
                float* data = (float*)outputs[i].data;
                for (int j = 0; j < outputs[i].rows; j++, data += outputs[i].cols)
                {
                    float confidence = data[4];
                    if (confidence > 0.5)  // 只考慮高信心度的偵測
                    {
                        int class_id = static_cast<int>(data[5]);
                        if (class_id == 0)  // YOLOv8 class 0 代表 "person"
                        {
                            int x = static_cast<int>(data[0] * image.cols);
                            int y = static_cast<int>(data[1] * image.rows);
                            int width = static_cast<int>(data[2] * image.cols);
                            int height = static_cast<int>(data[3] * image.rows);

                            detected_human_area = width * height;

                            // 在影像上畫出框
                            cv::rectangle(image, cv::Rect(x, y, width, height), cv::Scalar(0, 255, 0), 2);
                        }
                    }
                }
            }

            // 發布人員面積
            std_msgs::Float32 msg;
            msg.data = detected_human_area;
            human_area_pub_.publish(msg);

            // 顯示影像（可選）
            cv::imshow("Human Detection", image);
            cv::waitKey(1);
        }
        catch (cv_bridge::Exception& e)
        {
            ROS_ERROR("cv_bridge exception: %s", e.what());
        }
    }
};

int main(int argc, char** argv)
{
    ros::init(argc, argv, "human_detector");
    HumanDetector hd;
    ros::spin();
    return 0;
}

File: src/mobile/navigation_controller/src/human_detector.py
#!/usr/bin/env python3

import rospy
import torch
from ultralytics import YOLO
from sensor_msgs.msg import Image
from std_msgs.msg import Float32
from cv_bridge import CvBridge
import cv2

class HumanDetector:
    def __init__(self):
        rospy.init_node("human_detector", anonymous=True)

        # 訂閱相機影像
        self.image_sub = rospy.Subscriber("/camera/color/image_raw", Image, self.image_callback)

        # 設定發佈最大人員面積
        self.human_area_pub = rospy.Publisher("/human_detection_area", Float32, queue_size=10)

        # 讀取 YOLOv8 模型
        self.model = YOLO("yolov8m.pt")

        # OpenCV 影像轉換橋接
        self.bridge = CvBridge()

        # 存儲最新的最大人員面積
        self.latest_human_area = 0.0

        # 設置 0.5 秒定時器，每 0.5 秒執行一次 `publish_human_area`
        self.timer = rospy.Timer(rospy.Duration(0.5), self.publish_human_area)

        rospy.loginfo("HumanDetector Node Initialized")

    def image_callback(self, msg):
        try:
            # 轉換 ROS 影像到 OpenCV
            image = self.bridge.imgmsg_to_cv2(msg, "bgr8")
            image = cv2.resize(image, (640, 480))

            # 進行物件偵測
            results = self.model(image)

            max_human_area = 0.0  # 只存最大的人員面積

            for result in results:
                for box in result.boxes:
                    cls = int(box.cls[0].item())  # 物件類別
                    conf = box.conf[0].item()  # 信心度

                    if cls == 0 and conf > 0.7:  # 只偵測人 (YOLO class 0)
                        x1, y1, x2, y2 = map(int, box.xyxy[0].tolist())
                        width, height = x2 - x1, y2 - y1
                        area = width * height

                        if area > max_human_area:
                            max_human_area = area  # 只記錄最大面積

                        # 繪製框
                        cv2.rectangle(image, (x1, y1), (x2, y2), (0, 255, 0), 2)

            # 更新最新的最大人員面積
            self.latest_human_area = max_human_area

            # 顯示影像（可選）
            cv2.imshow("Human Detection", image)
            cv2.waitKey(1)

        except Exception as e:
            rospy.logerr(f"Error processing image: {e}")

    def publish_human_area(self, event):
        """定時發佈最大人員面積"""
        self.human_area_pub.publish(Float32(self.latest_human_area))
        rospy.loginfo(f"Published max human area: {self.latest_human_area}")

if __name__ == "__main__":
    HumanDetector()
    rospy.spin()
